internal/transport/grpchttp2/http2bridge_test.go

```diff
@@ -0,0 +1,609 @@
+	// to test merging.
+	c.rbuf = append(c.rbuf, 0, 0, byte(len(half2)), byte(FrameTypeContinuation), byte(FlagContinuationEndHeaders))
+	c.rbuf = appendUint32(c.rbuf, 1)
+	// Copy data written by the encoder into the reading buf
+	c.rbuf = append(c.rbuf, half2...)
+
+	f := NewFramerBridge(c, c, 0, nil)
+	fr, err := f.ReadFrame()
+	if err != nil {
+		t.Fatalf("ReadFrame(): %v", err)
+	}
+
+	mf, ok := fr.(*MetaHeadersFrame)
+	if !ok {
+		t.Errorf("ReadFrame(): Type: expected MetaHeadersFrame, got %T", fr)
+	}
+	if len(mf.Fields) != 2 {
+		t.Errorf("ReadFrame(): Fields: got %d, want %d", len(mf.Fields), 1)
+	}
+	for i, field := range fields {
+		if field.Name != mf.Fields[i].Name {
+			t.Errorf("ReadFrame(): Fields[%d].Name: got %q, want %q", i, mf.Fields[i].Name, field.Name)
+		}
+		if field.Value != mf.Fields[i].Value {
+			t.Errorf("ReadFrame(): Fields[%d].Value: got %q, want %q", i, mf.Fields[i].Value, field.Value)
+		}
+	}
+}
+
+// Tests and verifies that a Data Frame is correctly written.
+func (s) TestBridge_WriteData(t *testing.T) {
+	c := &testConn{}
+	wantData := "test data"
+	testBuf := [][]byte{[]byte("test"), []byte(" data")}
+	f := NewFramerBridge(c, c, 0, nil)
+	f.WriteData(1, false, testBuf...)
+	wantHdr := FrameHeader{
+		Size:     uint32(len(wantData)),
+		Type:     FrameTypeData,
+		Flags:    0,
+		StreamID: 1,
+	}
+	if errs := checkWrittenHeader(c.wbuf[0:9], wantHdr); len(errs) > 0 {
+		t.Errorf("WriteData():\n%s", errors.Join(errs...))
+	}
+	if string(c.wbuf[9:]) != wantData {
+		t.Errorf("WriteData(): Data: got %q, want %q", string(c.wbuf[9:]), wantData)
+	}
+}
+
+// Tests and verifies that a Headers Frame and all its flag permutations are
+// correctly written.
+func (s) TestBridge_WriteHeaders(t *testing.T) {
+	tests := []struct {
+		name       string
+		endStream  bool
+		endHeaders bool
+	}{
+		{name: "no flags", endStream: false, endHeaders: false},
+		{name: "endheaders", endStream: false, endHeaders: true},
+		{name: "endstream and endheaders", endStream: true, endHeaders: true},
+		{name: "endstream", endStream: true, endHeaders: false},
+	}
+	wantData := "test data"
+
+	for _, test := range tests {
+		t.Run(fmt.Sprintf(test.name, test.endStream, test.endHeaders), func(t *testing.T) {
+			c := &testConn{}
+			f := NewFramerBridge(c, c, 0, nil)
+
+			f.WriteHeaders(1, test.endStream, test.endHeaders, []byte(wantData))
+
+			var wantFlags Flag
+			if test.endStream {
+				wantFlags |= FlagHeadersEndStream
+			}
+			if test.endHeaders {
+				wantFlags |= FlagHeadersEndHeaders
+			}
+			wantHdr := FrameHeader{
+				Size:     uint32(len(wantData)),
+				Type:     FrameTypeHeaders,
+				Flags:    wantFlags,
+				StreamID: 1,
+			}
+			if errs := checkWrittenHeader(c.wbuf[0:9], wantHdr); len(errs) > 0 {
+				t.Errorf("WriteHeaders():\n%s", errors.Join(errs...))
+			}
+			if data := string(c.wbuf[9:]); data != wantData {
+				t.Errorf("WriteHeaders(): Data: got %q, want %q", data, wantData)
+			}
+			c.wbuf = c.wbuf[:0]
```

Comment:
what's this for?

Reply:
Oh right. This was before when the for loop did not use `t.Run`, so all permutations shared the same test conn. I'll remove it.